File: frontend/web/src/services/auth-service.ts
/**
 * Zero-Knowledge Authentication Service
 * 
 * Integrates CryptoService with backend authentication endpoints.
 * Handles client-side key derivation and DEK management.
 * 
 * Flow:
 * - Signup: Generate DEK, derive Master Key, encrypt DEK, send to backend
 * - Login: Retrieve encrypted DEK, derive Master Key, decrypt DEK, store locally
 * - Logout: Clear DEK from local storage
 */

import { cryptoService } from "./crypto-service";
import { secureStorage } from "./secure-storage";

// Configuration
const API_BASE_URL = process.env.REACT_APP_API_URL || "https://api.photosden.com/dev";

interface SignupRequest {
    email: string;
    password: string;
}

interface SignupResponse {
    verificationRequired: boolean;
    userId: string;
    correlationId: string;
}

interface LoginRequest {
    email: string;
    password: string;
}

interface LoginResponse {
    idToken: string;
    accessToken: string;
    refreshToken: string;
    expiresIn: number;
    userId: string;
    
    // Encryption parameters
    encryptedDEK: string;
    kdfSalt: string;
    kdfIterations: number;
    kdfAlgorithm: string;
    
    correlationId: string;
}

interface PasswordChangeRequest {
    oldPassword: string;
    newPassword: string;
}

export class AuthService {
    private currentUserId: string | null = null;
    private accessToken: string | null = null;

    /**
     * Signup with zero-knowledge encryption
     * 
     * Steps:
     * 1. Generate random DEK
     * 2. Generate random salt
     * 3. Derive Master Key from password
     * 4. Encrypt DEK with Master Key
     * 5. Send encrypted DEK + salt to backend
     * 6. Store DEK locally (only after successful signup)
     */
    async signup(request: SignupRequest): Promise<SignupResponse> {
        try {
            // Step 1: Generate DEK (this will encrypt ALL user files)
            const dek = await cryptoService.generateDEK();

            // Step 2: Generate random salt for PBKDF2
            const salt = cryptoService.generateSalt();

            // Step 3: Derive Master Key from password
            const masterKey = await cryptoService.deriveMasterKey(
                request.password,
                salt
            );

            // Step 4: Encrypt DEK with Master Key
            const { encryptedDEK } = await cryptoService.encryptDEK(dek, masterKey);

            // Step 5: Send to backend
            const response = await fetch(`${API_BASE_URL}/auth/signup`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    identifier: request.email,
                    password: request.password,
                    encryptedDEK,
                    kdfSalt: cryptoService.arrayBufferToBase64(salt),
                    kdfIterations: 100_000,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.message || "Signup failed");
            }

            const data: SignupResponse = await response.json();

            // Step 6: Store DEK locally for future use
            // NOTE: Only store AFTER successful backend response
            await secureStorage.storeDEK(data.userId, dek);

            return data;
        } catch (error: any) {
            console.error("Signup error:", error);
            throw new Error(`Signup failed: ${error.message}`);
        }
    }

    /**
     * Login with zero-knowledge encryption
     * 
     * Steps:
     * 1. Authenticate with backend (get encrypted DEK + salt)
     * 2. Derive Master Key from password
     * 3. Decrypt DEK with Master Key
     * 4. Store DEK locally for session
     */
    async login(request: LoginRequest): Promise<LoginResponse> {
        try {
            // Step 1: Authenticate with backend
            const response = await fetch(`${API_BASE_URL}/auth/login`, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                },
                body: JSON.stringify({
                    identifier: request.email,
                    password: request.password,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.message || "Login failed");
            }

            const data: LoginResponse = await response.json();

            // Step 2: Derive Master Key from password
            const salt = this.base64ToUint8Array(data.kdfSalt);
            const masterKey = await cryptoService.deriveMasterKey(
                request.password,
                salt,
                data.kdfIterations
            );

            // Step 3: Decrypt DEK with Master Key
            const dek = await cryptoService.decryptDEK(
                data.encryptedDEK,
                masterKey
            );

            // Step 4: Store DEK locally
            await secureStorage.storeDEK(data.userId, dek);

            // Store tokens for API calls
            this.currentUserId = data.userId;
            this.accessToken = data.accessToken;

            return data;
        } catch (error: any) {
            console.error("Login error:", error);
            throw new Error(`Login failed: ${error.message}`);
        }
    }

    /**
     * Change password with DEK re-encryption
     * 
     * Steps:
     * 1. Get current DEK from local storage
     * 2. Generate new salt
     * 3. Derive NEW Master Key from new password
     * 4. Re-encrypt SAME DEK with NEW Master Key
     * 5. Send to backend
     * 
     * CRITICAL: DEK never changes, only its encryption wrapper
     */
    async changePassword(request: PasswordChangeRequest): Promise<void> {
        if (!this.currentUserId || !this.accessToken) {
            throw new Error("Not authenticated");
        }

        try {
            // Step 1: Get current DEK
            const dek = await secureStorage.getDEK(this.currentUserId);
            if (!dek) {
                throw new Error("DEK not found. Please login again.");
            }

            // Step 2: Generate new salt
            const newSalt = cryptoService.generateSalt();

            // Step 3: Derive NEW Master Key from new password
            const newMasterKey = await cryptoService.deriveMasterKey(
                request.newPassword,
                newSalt
            );

            // Step 4: Re-encrypt SAME DEK with NEW Master Key
            const { encryptedDEK: newEncryptedDEK } = await cryptoService.encryptDEK(
                dek,
                newMasterKey
            );

            // Step 5: Send to backend
            const response = await fetch(`${API_BASE_URL}/profile/password`, {
                method: "PUT",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${this.accessToken}`,
                },
                body: JSON.stringify({
                    oldPassword: request.oldPassword,
                    newPassword: request.newPassword,
                    newEncryptedDEK,
                    newKdfSalt: cryptoService.arrayBufferToBase64(newSalt),
                    newKdfIterations: 100_000,
                }),
            });

            if (!response.ok) {
                const error = await response.json();
                
                // Handle critical partial failure
                if (error.error === "PartialFailure") {
                    throw new Error(
                        "CRITICAL: Password changed but encryption update failed. Contact support immediately!  UserId: " +
                            error.userId
                    );
                }
                
                throw new Error(error.message || "Password change failed");
            }

            // Success - DEK remains the same, just re-encrypted
        } catch (error: any) {
            console.error("Password change error:", error);
            throw error;
        }
    }

    /**
     * Logout - clear DEK from local storage
     */
    async logout(): Promise<void> {
        if (this.currentUserId) {
            await secureStorage.clearDEK(this.currentUserId);
        }

        this.currentUserId = null;
        this.accessToken = null;
    }

    /**
     * Check if user has valid session (DEK available)
     */
    async hasValidSession(userId: string): Promise<boolean> {
        return await secureStorage.hasDEK(userId);
    }

    /**
     * Get current user ID
     */
    getCurrentUserId(): string | null {
        return this.currentUserId;
    }

    /**
     * Get access token for API calls
     */
    getAccessToken(): string | null {
        return this.accessToken;
    }

    /**
     * Set access token (for page refresh scenarios)
     */
    setAccessToken(token: string, userId: string): void {
        this.accessToken = token;
        this.currentUserId = userId;
    }

    /**
     * Helper: Convert Base64 to Uint8Array
     */
    private base64ToUint8Array(base64: string): Uint8Array {
        const binary = atob(base64);
        const bytes = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            bytes[i] = binary.charCodeAt(i);
        }
        return bytes;
    }
}

// Export singleton instance
export const authService = new AuthService();

File: mobile/lib/core/services/auth_service.dart
import 'dart:convert';
import 'package:cryptography/cryptography.dart';
import 'package:http/http.dart' as http;
import '../services/crypto_service.dart';
import '../services/secure_storage_service.dart';

/// Zero-Knowledge Authentication Service
///
/// Integrates CryptoService with backend authentication endpoints.
/// Handles client-side key derivation and DEK management.
///
/// Flow:
/// - Signup: Generate DEK, derive Master Key, encrypt DEK, send to backend
/// - Login: Retrieve encrypted DEK, derive Master Key, decrypt DEK, store locally
/// - Logout: Clear DEK from local storage
class AuthService {
  final CryptoService _cryptoService;
  final SecureStorageService _storageService;
  final String  _apiBaseUrl;

  String? _accessToken;
  String? _currentUserId;

  AuthService({
    required CryptoService cryptoService,
    required SecureStorageService storageService,
    required String apiBaseUrl,
  })  : _cryptoService = cryptoService,
        _storageService = storageService,
        _apiBaseUrl = apiBaseUrl;

  /// Signup with zero-knowledge encryption
  ///
  /// Steps:
  /// 1. Generate random DEK
  /// 2. Generate random salt
  /// 3. Derive Master Key from password
  /// 4. Encrypt DEK with Master Key
  /// 5. Send encrypted DEK + salt to backend
  /// 6. Store DEK locally (only after successful signup)
  Future<SignupResponse> signup({
    required String email,
    required String password,
  }) async {
    try {
      // Step 1: Generate DEK (this will encrypt ALL user files)
      final dek = await _cryptoService.generateDEK();

      // Step 2: Generate random salt for PBKDF2
      final salt = _cryptoService.generateSalt();

      // Step 3: Derive Master Key from password
      final masterKey = await _cryptoService.deriveMasterKey(password, salt);

      // Step 4: Encrypt DEK with Master Key
      final encryptedDEKResult = await _cryptoService.encryptDEK(dek, masterKey);

      // Step 5: Send to backend
      final response = await http.post(
        Uri.parse('$_apiBaseUrl/auth/signup'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'identifier': email,
          'password': password,
          'encryptedDEK': encryptedDEKResult.encryptedDEK,
          'kdfSalt': base64Encode(salt),
          'kdfIterations': 100000,
        }),
      );

      if (response.statusCode != 200 && response.statusCode != 201) {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Signup failed');
      }

      final data = jsonDecode(response.body);

      // Step 6: Store DEK locally for future use
      final userId = data['userId'] as String;
      await _storageService.storeDEK(userId, dek);

      return SignupResponse(
        verificationRequired: data['verificationRequired'] ?? true,
        userId: userId,
        correlationId: data['correlationId'] ?? '',
      );
    } catch (e) {
      print('Signup error: $e');
      throw Exception('Signup failed: $e');
    }
  }

  /// Login with zero-knowledge encryption
  ///
  /// Steps:
  /// 1. Authenticate with backend (get encrypted DEK + salt)
  /// 2. Derive Master Key from password
  /// 3. Decrypt DEK with Master Key
  /// 4. Store DEK locally for session
  Future<LoginResponse> login({
    required String email,
    required String password,
  }) async {
    try {
      // Step 1: Authenticate with backend
      final response = await http.post(
        Uri.parse('$_apiBaseUrl/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'identifier': email,
          'password': password,
        }),
      );

      if (response.statusCode != 200) {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Login failed');
      }

      final data = jsonDecode(response.body);

      // Step 2: Derive Master Key from password
      final salt = base64Decode(data['kdfSalt']);
      final iterations = data['kdfIterations'] as int;
      final masterKey = await _cryptoService.deriveMasterKey(
        password,
        salt,
        iterations: iterations,
      );

      // Step 3: Decrypt DEK with Master Key
      final dek = await _cryptoService.decryptDEK(
        data['encryptedDEK'],
        masterKey,
      );

      // Step 4: Store DEK locally
      final userId = data['userId'] as String;
      await _storageService.storeDEK(userId, dek);

      // Store tokens for API calls
      _currentUserId = userId;
      _accessToken = data['accessToken'];

      return LoginResponse(
        idToken: data['idToken'],
        accessToken: data['accessToken'],
        refreshToken: data['refreshToken'],
        expiresIn: data['expiresIn'],
        userId: userId,
        correlationId: data['correlationId'] ?? '',
      );
    } catch (e) {
      print('Login error: $e');
      throw Exception('Login failed: $e');
    }
  }

  /// Change password with DEK re-encryption
  ///
  /// Steps:
  /// 1. Get current DEK from local storage
  /// 2. Generate new salt
  /// 3. Derive NEW Master Key from new password
  /// 4. Re-encrypt SAME DEK with NEW Master Key
  /// 5. Send to backend
  ///
  /// CRITICAL: DEK never changes, only its encryption wrapper
  Future<void> changePassword({
    required String oldPassword,
    required String newPassword,
  }) async {
    if (_currentUserId == null || _accessToken == null) {
      throw Exception('Not authenticated');
    }

    try {
      // Step 1: Get current DEK
      final dek = await _storageService.getDEK(_currentUserId!);
      if (dek == null) {
        throw Exception('DEK not found. Please login again.');
      }

      // Step 2: Generate new salt
      final newSalt = _cryptoService.generateSalt();

      // Step 3: Derive NEW Master Key from new password
      final newMasterKey = await _cryptoService.deriveMasterKey(
        newPassword,
        newSalt,
      );

      // Step 4: Re-encrypt SAME DEK with NEW Master Key
      final newEncryptedDEK = await _cryptoService.encryptDEK(dek, newMasterKey);

      // Step 5: Send to backend
      final response = await http.put(
        Uri.parse('$_apiBaseUrl/profile/password'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $_accessToken',
        },
        body: jsonEncode({
          'oldPassword': oldPassword,
          'newPassword': newPassword,
          'newEncryptedDEK': newEncryptedDEK.encryptedDEK,
          'newKdfSalt': base64Encode(newSalt),
          'newKdfIterations': 100000,
        }),
      );

      if (response.statusCode != 200) {
        final error = jsonDecode(response.body);

        // Handle critical partial failure
        if (error['error'] == 'PartialFailure') {
          throw Exception(
            'CRITICAL: Password changed but encryption update failed. '
            'Contact support immediately! UserId: ${error['userId']}',
          );
        }

        throw Exception(error['message'] ?? 'Password change failed');
      }

      // Success - DEK remains the same, just re-encrypted
    } catch (e) {
      print('Password change error: $e');
      rethrow;
    }
  }

  /// Logout - clear DEK from local storage
  Future<void> logout() async {
    if (_currentUserId != null) {
      await _storageService.clearDEK(_currentUserId!);
    }

    _currentUserId = null;
    _accessToken = null;
  }

  /// Check if user has valid session (DEK available)
  Future<bool> hasValidSession(String userId) async {
    return await _storageService.hasDEK(userId);
  }

  /// Get current user ID
  String? get currentUserId => _currentUserId;

  /// Get access token for API calls
  String? get accessToken => _accessToken;

  /// Set access token (for page refresh scenarios)
  void setAccessToken(String token, String userId) {
    _accessToken = token;
    _currentUserId = userId;
  }
}

/// Signup response
class SignupResponse {
  final bool verificationRequired;
  final String userId;
  final String correlationId;

  SignupResponse({
    required this.verificationRequired,
    required this.userId,
    required this.correlationId,
  });
}

/// Login response
class LoginResponse {
  final String idToken;
  final String accessToken;
  final String refreshToken;
  final int expiresIn;
  final String userId;
  final String correlationId;

  LoginResponse({
    required this.idToken,
    required this.accessToken,
    required this.refreshToken,
    required this.expiresIn,
    required this.userId,
    required this.correlationId,
  });
}
